### PR TITLE
gh-84995: Run sys.__interactivehook__() on asyncio REPL startup

### PR DESCRIPTION
**Pull request by remilapeyre**
https://bugs.python.org/issue40818

* Issue: gh-84995

### CODE REVIEW

File: Lib/asyncio/__main__.py [hunk 1]
```diff
@@ -109,6 +110,24 @@ def run(self):
     except ImportError:
         pass
 
+    try:
+        interactive_hook = sys.__interactivehook__
+    except AttributeError():
```
Comment (ZackerySpytz) [inline Lib/asyncio/__main__.py]: I don't think `()` should be here.

File: Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+The asyncio REPL now run :func:`sys.__interactivehook__` on startup. The
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst]: The docs use the `data` role for `sys.__interactivehook__`, not `func`.

File: Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+default implementation of :func:`sys.__interactivehook__` gives
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst]: Ditto.

**Comment by asvetlov** [comment]
@1st1 could you review?

**Comment by andriisoldatenko** [comment]
@1st1 is anything missing in this PR?

**Comment by willingc** [comment]
Hi @remilapeyre, Is this PR something you still wish to have open? If you would like to move it forward, we would need a signed CLA and merge conflicts resolved. Please close if you do not wish to continue working on this. Thanks.

**Review by willingc — COMMENTED** [review]
Please sign CLA and resolve merge conflict. Thanks.

**Comment by gvanrossum** [comment]
Looks like OP lost interest (or was never notified). I'll add "Pending" label, if there's no response in 30 days just close the PR.

**Comment by gvanrossum** [comment]
@remilapeyre Please fix merge conflict.

**Comment by remilapeyre** [comment]
Hello, I'm a bit short on time to contribute but will try to put aside some time in the coming weeks to revive the PRs

**Comment by gvanrossum** [comment]
@willingc Can i leave this to you?

**Comment by itamaro** [comment]
gh-116095 is a "re-do" of this PR on top of main (sorry, I don't have access to update this PR directly, but I have credited @remilapeyre as the original author)

File: Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+The asyncio REPL now run :data:`sys.__interactivehook__` on startup. The
+default implementation of :data:`sys.__interactivehook__` gives
+auto-completion to the asyncio REPL. Patch contributed by Rémi Lapeyre.
```
Comment (itamaro) [inline Misc/NEWS.d/next/Library/2020-05-29-18-08-54.bpo-40818.Ij8ffq.rst]:
```suggestion
The asyncio REPL now runs :data:`sys.__interactivehook__` on startup. The
default implementation of :data:`sys.__interactivehook__` provides
auto-completion to the asyncio REPL. Patch contributed by Rémi Lapeyre.
```

File: Lib/site.py [hunk 5]
```diff
@@ -457,63 +457,58 @@ def gethistoryfile():
         '.python_history')
 
 
-def enablerlcompleter():
-    """Enable default readline configuration on interactive prompts, by
-    registering a sys.__interactivehook__.
+def register_readline():
+    """Enable default readline configuration on interactive prompts.
```
Comment (gvanrossum) [inline Lib/site.py]: Technically, this function *is* the "interactive hook". The docstring should probably clarify that.

Also, even though it isn't documented, I think we should maintain a function named `enablerlcompleter()` which, when calls, does
```py
sys.__interactivehook__ = register_readline
```
It's likely that someone, somewhere is relying on that function existing, and it's only 2-3 lines to keep them happy.

File: Lib/site.py [hunk 6]
```diff
@@ -457,63 +457,58 @@ def gethistoryfile():
...
+    except OSError:
+        # An OSError here could have many causes, but the most likely one
+        # is that there's no .inputrc file (or .editrc file in the case of
+        # Mac OS X + libedit) in the expected location.  In that case, we
+        # want to ignore the exception.
+        pass
+
+    if readline.get_current_history_length() == 0:
+        # If no history was loaded, default to .python_history, or
+        # PYTHON_HISTORY.
```
Comment (gvanrossum) [inline Lib/site.py]: I'd keep the original formatting:
```suggestion
        # If no history was loaded, default to .python_history,
        # or PYTHON_HISTORY.
```

File: Lib/site.py [hunk 7]
```diff
@@ -457,63 +457,58 @@ def gethistoryfile():
...
+        # The guard is necessary to avoid doubling history size at
+        # each interpreter exit when readline was already configured
+        # through a PYTHONSTARTUP hook, see:
+        # http://bugs.python.org/issue5845#msg198636
         try:
-            readline.read_init_file()
+            readline.read_history_file(gethistoryfile())
```
Comment (gvanrossum) [inline Lib/site.py]: I'd prefer to keep the original code, which calls `gethistoryfile()` only once. Calling it a second time changes the semantics (someone might have changed `os.environ["PYTHON_HISTORY"]` or `HOME`).

Call me chicken, but let's not change more than absolutely necessary to implement your feature.

File: Lib/asyncio/__main__.py [hunk 8]
```diff
@@ -109,6 +110,24 @@ def run(self):
     except ImportError:
         pass
 
+    try:
+        interactive_hook = sys.__interactivehook__
+    except AttributeError:
+        pass
+    else:
+        if interactive_hook is not None:
```
Comment (gvanrossum) [inline Lib/asyncio/__main__.py]: You could do without the `try` / `except` / `else` by using `interactive_hook = getattr(sys, "__interactivehook__", None)`.

**Review by gvanrossum — COMMENTED** [review]
I like the feature and I like the implementation, but I have a few suggestions to reduce the risk of accidentally breaking code that was relying on implementation details of site.py. (Note that I am fine with moving `register_readline()` to the toplevel scope in site.py.)

File: Lib/site.py (hunk 7 above)
Comment (remilapeyre) [reply Lib/site.py]: Thanks, I botched this while fixing the conflicts

File: Lib/site.py (hunk 5 above)
Comment (remilapeyre) [reply Lib/site.py]: I added `enablerlcompleter()` back and changed the wording in the docstrings. Should I add `enablerlcompleter()` and `register_readline()` to the documentation?

**Review by gvanrossum — APPROVED** [review]
Thanks! I'll merge it now. I wouldn't add more docs about this to site.rst, it's intentionally a bit vague.

**Pull request merged**